New running mod

kevingranade opened the pull request:
Built on top of Reaper's stamina/running mod.
The general outline is:
You can toggle on running, at full stamina it doubles your move speed (technically halves move cost) at the cost of burning through your stamina rapidly.  The speed boost decays as your stamina drops, until it disappears entirely when your stamina reaches 0, at which point you're exhausted, and moving even slower than normal walking speed.  Walking speed also decays with stamina loss, but you slowly recover stamina while walking (unless your mouth is heavily encumbered!).
The stamina drain scales with your encumbrance level, so carrying a lot of stuff makes you get tired faster.
You regain stamina slowly while walking, and rapidly while inactive.
Throwing things, firing bows, and using melee weapons also drains stamina, and if your stamina is low (under 25% currently) they gain move penalties.
Adapted the winded effect, it causes a few debuffs (very open to feedback on them), but more importantly blocks stamina recovery while it's in effect.  Continuing to run while you're winded is really bad news as it will drain your stamina much faster.

TODO:
- [x] Adapt winded disease from #8888 for when your stamina becomes critically low.
- [ ] Add more actions that drain stamina.
  - [ ] Dragging vehicles
  - [x] Smashing things
  - [x] Riding bicycles.
- [ ] Integrate with diseases and mutations.
- [x] Make reaching 0 stamina automatically turn off running.
- [x] Add move mode indicator to UI.

AllisonBW commented:
I don't mean to necro, but a thought: how about tying stamina to the (currently) hidden Health stat? It's already in the game, improvable, and logically and realistically something that would influence stamina. Also, it might make healthy living slightly more relevant to the people who just wear permanent filter masks to dodge colds.

Would love to see this feature in, by the by. I think it'd dovetail especially well with the horde fixes I heard Kevin wanted to work on, given that hordes would make fleeing from danger a bigger part of gameplay beyond the early game.

Zireael07 commented:
I'd love it if it made it in.

kevingranade commented:
Having health be an aspect of it sounds reasonable.
I'd also love to get this in, it's the next big feature I'm going to work
on, but I've had no time recently, and right now I'm trying to push in the
direction of a release.

kevingranade commented:
Back on this obviously.  Balance is still bad, you can take a significant amount of pain and easily get away from a zombie, and even have a ton of stamina left over, and that's on open ground.  In a city it would make getting away quite trivial.

Rivet-the-Zombie commented:
I see the TODO list is still fairly empty. Is this ready for merge testing?

kevingranade commented:
Nope, I think this predates us using the wip tag consistently.

Lain- commented:
Nice to see this going again.

Rivet-the-Zombie commented:
Progress tags are our friends.

KA101 commented on src/player.cpp:
```diff
@@ -12942,6 +12976,23 @@ int player::get_hp_max( hp_part bp ) const
 
+int player::get_stamina_max()
+{
+    return 1000;
```
Hmm.  I'd like to have stronger folks have more endurance, since we don't differentiate that out of ST.  (And it would make DX less the UberStat for melee purposes.)

Rivet-the-Zombie replied:
I second that.

Snaaty replied:
Is a new stat completely out of scope for this PR? 
If so, buffing dex further is probably not necessary.

Lain- replied:
A STA (Stamina) stat wouldn't so bad I think, to differentiate from STR (the pure brute force you can apply) and DEX (how dextrous you are). Though that would require a complex rebalance of many things...

Zireael07 replied:
I'm against a new stat.
Tie stamina to ST or DX or maybe a bit to both.

kevingranade replied:
See previous discussion, I don't like tying it to strength, and it's out of scope for this PR IMO to add a new stat.
I want to get some feedback on how it plays before adding more complications like that.
The main thing holding it is that as it is now it's simply a huge buff to escaping from enemies as it makes the player strictly faster, and I'm experimenting with different things to compensate for that.

KA101 replied:
OK.  I was thinking more along the lines of being able to swing that melee weapon around more, not necessarily running, but that means touching the process in a different area.  Thanks.

kevingranade replied:
I'm of two minds with strength and stamina loss from melee.  On one hand, it makes sense that stronger characters aren't exerting as much.  On the other hand, they're already getting more damage per swing, doesn't that mean they're exerting themselves more?

Rivet-the-Zombie replied:
Or they're exerting themselves a proportionately equal amount, but with a greater result.

kevingranade commented:
Rebased, added several more features I was going for, I might poke at it a bit, but at this point I'm not sure the remaining features are worth keeping it out of mainline.

DavidKeaton commented:
It must be CHRISTMAS.

narc0tiq commented:
April 1 is actually the official Internet Christmas, yes.

On Wed, Apr 1, 2015 at 3:15 PM Davek <notifications@github.com> wrote:

> It must be CHRISTMAS.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/9014#issuecomment-88455505>
> .
>

DavidKeaton commented:
I knew it.
